aten/src/ATen/core/TensorMethods.h

```diff
@@ -1240,6 +1231,44 @@ inline Device Tensor::device() const {
   return Device(type().device_type(), type().is_cuda() ? get_device() : -1);
 }
 
+inline int64_t Tensor::get_device() const {
+  // NB: This gets called a lot so we special case it here instead of dispatching
+  // to a native function.
+  const auto& tid = impl_->type_id();
+  if (tid == CUDATensorId()) {
+    // TODO: #12934 Investigate caching device on TensorImpl for performance
+    return impl_->storage().device().index();
+  } else if (tid == SparseCUDATensorId()) {
+    return _values().get_device();
+  } else {
+    AT_ERROR("get_device is not implemented for type ", type());
+  }
+}
+
+inline int64_t get_device(Tensor self) {
+  return self.get_device();
+}
+
+inline bool Tensor::is_cuda() const {
+  // NB: avoids dispatch for perf reasons
+  const auto& tid = impl_->type_id();
+  return tid == CUDATensorId() || tid == SparseCUDATensorId();
+}
+
+inline bool is_cuda(Tensor self) {
+  return self.is_cuda();
+}
+
+inline bool Tensor::is_sparse() const {
```

Comment:
This method would mean that sparse is no longer open, which could be problematic in practice because one could very reasonably want to add a `SparseHIPTensorId`, e.g.

If we are going to allow people to write code like this, it seems like a strong argument to make TensorTypeId as a closed enum, and then implement all of these "attribute" tests in the file (so we remember to update them when we add other TensorTypeIds).

If we *don't* want to allow this, then we had better have another, non-vdispatched way to check what the memory layout of a tensor is. Maybe some sort of "Layout" field in TensorImpl?

Reply:
I think we should just dodge this question and force the implementation into TensorTypeId.  I don't have a strong preference for returning a bool or a layout enum or whatever for now.  But this isn't maintainable.

Reply:
OK. @zou3519, just to make the ask clear: we'd like this implementation to live in `TensorImpl`, rather than as a method on `Tensor`. API doesn't have to change; we just want this to live closer to where we could make implementation changes to change how we implement this.

Reply:
Sounds good, thanks for the clarification!

Reply:
Do you want all of `{ is_sparse, is_cuda, get_device }` to live in `TensorImpl`?

Reply:
Also, @ezyang, @gchanan, is it still okay to special case these functions in `Tensor`? They would end up calling the relevant methods on TensorImpl.

I'm trying to avoid a type() dispatch because those are costly (8ns) compared to how fast `is_cuda` and `is_sparse` can theoretically be.

Reply:
It's OK to special case them on tensor; just have them call the method on TensorImpl.